core/src/main/java/hudson/util/io/TarArchiver.java

```diff
@@ -87,27 +90,46 @@ public void visit(File file, String relativePath) throws IOException {
         int mode = IOUtils.mode(file);
         if (mode!=-1)   te.setMode(mode);
         te.setModTime(file.lastModified());
-        if(!file.isDirectory())
-            te.setSize(file.length());
-
-        tar.putArchiveEntry(te);
+        long size = 0;
 
         if (!file.isDirectory()) {
-            FileInputStream in = new FileInputStream(file);
-            try {
-                int len;
-                while((len=in.read(buf))>=0)
-                    tar.write(buf,0,len);
-            } finally {
-                in.close();
+            size = file.length();
+            te.setSize(size);
+        }
+        tar.putArchiveEntry(te);
+        try {
+            if (!file.isDirectory()) {
+                FileInputStream in = new FileInputStream(file);
+                try {
+                    int len;
+                    int read = 0;
+                    while ((len = in.read(buf)) >= 0) {
+                        if (len + read <= size) { // ensure we don't write more bytes than the declared when we created
+                                                  // the entry
+                            tar.write(buf, 0, len);
+                            read += len;
+                        } else {
+                            tar.write(buf, 0, (int) size - read);
+                            break;
+                        }
+                    }
+                } catch (IOException e) {// log the exception in any case
+                    LOGGER.log(Level.SEVERE, "Error writing to tar file from: " + file, e);
+                    throw e;
+                } finally {
+                    in.close();
+                }
             }
+        } finally { // always close the entry
```

Comment:
I do not think this needs to be in a `finally`-block. Either there was an error, in which case the tarball is trash, or there was not, in which case it is closed in the normal course of things. The tarball output stream is closed elsewhere I suppose.